pkg/logtailing/tailer.go

```diff
@@ -0,0 +1,213 @@
+
+	"github.com/stripe/stripe-cli/pkg/ansi"
+	"github.com/stripe/stripe-cli/pkg/stripeauth"
+	"github.com/stripe/stripe-cli/pkg/websocket"
+)
+
+const outputFormatJSON = "json"
+
+// LogFilters contains all of the potential user-provided filters for log tailing
+type LogFilters struct {
+	FilterAccount        []string `json:"filter_account,omitempty"`
+	FilterIPAddress      []string `json:"filter_ip_address,omitempty"`
+	FilterHTTPMethod     []string `json:"filter_http_method,omitempty"`
+	FilterRequestPath    []string `json:"filter_request_path,omitempty"`
+	FilterRequestStatus  []string `json:"filter_request_status,omitempty"`
+	FilterSource         []string `json:"filter_source,omitempty"`
+	FilterStatusCode     []string `json:"filter_status_code,omitempty"`
+	FilterStatusCodeType []string `json:"filter_status_code_type,omitempty"`
+}
+
+// Config provides the configuration of a log tailer
+type Config struct {
+	APIBaseURL string
+
+	// DeviceName is the name of the device sent to Stripe to help identify the device
+	DeviceName string
+
+	// Filters for API request logs
+	Filters *LogFilters
+
+	// Key is the API key used to authenticate with Stripe
+	Key string
+
+	// Info, error, etc. logger. Unrelated to API request logs.
+	Log *log.Logger
+
+	// Force use of unencrypted ws:// protocol instead of wss://
+	NoWSS bool
+
+	// Output format for request logs
+	OutputFormat string
+
+	// WebSocketFeature is the feature specified for the websocket connection
+	WebSocketFeature string
+}
+
+// Tailer is the main interface for running the log tailing session
+type Tailer struct {
+	cfg *Config
+
+	stripeAuthClient *stripeauth.Client
+	webSocketClient  *websocket.Client
+
+	interruptCh chan os.Signal
+}
+
+// EventPayload is the mapping for fields in event payloads from request log tailing
+type EventPayload struct {
+	CreatedAt string `json:"created_at"`
+	Method    string `json:"method"`
+	RequestID string `json:"request_id"`
+	Status    int    `json:"status"`
+	URL       string `json:"url"`
+}
+
+// New creates a new Tailer
+func New(cfg *Config) *Tailer {
+	if cfg.Log == nil {
+		cfg.Log = &log.Logger{Out: ioutil.Discard}
+	}
+	return &Tailer{
+		cfg: cfg,
+		stripeAuthClient: stripeauth.NewClient(cfg.Key, &stripeauth.Config{
+			Log:        cfg.Log,
+			APIBaseURL: cfg.APIBaseURL,
+		}),
+		interruptCh: make(chan os.Signal, 1),
+	}
+}
+
+// Run sets the websocket connection
+func (tailer *Tailer) Run() error {
+	s := ansi.StartSpinner("Getting ready...", tailer.cfg.Log.Out)
+
+	// Intercept Ctrl+c so we can do some clean up
+	signal.Notify(tailer.interruptCh, os.Interrupt, syscall.SIGTERM)
+
+	filters, err := jsonifyFilters(tailer.cfg.Filters)
+	if err != nil {
+		tailer.cfg.Log.Fatalf("Error while converting log filters to JSON encoding: %v", err)
+	}
+
+	session, err := tailer.stripeAuthClient.Authorize(tailer.cfg.DeviceName, tailer.cfg.WebSocketFeature, &filters)
+	if err != nil {
+		// TODO: better error handling / retries
```

Comment:
Can't remember if I asked, but anything we should do before we merge? If not, could you create a ticket to track work here?

Reply:
I don't think so. This was also commented in the corresponding location in `proxy.go` so I just copied it over here as well. I'm not 100% sure what it was commented for originally, perhaps just auth retrying, but we can remove it if desired since it doesn't really seem to be an action item right now.